File: AutoEmply/Services/FormStructureValidator.cs
using AutoEmply.Models;

namespace AutoEmply.Services;

public static class FormStructureValidator
{
    /// <summary>
    /// Validates the FormStructure. Only returns hard errors that cannot be auto-corrected.
    /// Soft issues (colSpan mismatch, fraction sum) are auto-fixed by NormalizeInPlace.
    /// </summary>
    public static List<string> Validate(FormStructure? structure)
    {
        var errors = new List<string>();

        if (structure is null)
        {
            errors.Add("FormStructure is null.");
            return errors;
        }

        if (structure.Sections.Count == 0)
        {
            errors.Add("FormStructure must have at least one section.");
            return errors;
        }

        // Auto-fix soft issues before validation
        NormalizeInPlace(structure);

        for (var i = 0; i < structure.Sections.Count; i++)
        {
            ValidateSection(structure.Sections[i], $"sections[{i}]", errors);
        }

        return errors;
    }

    /// <summary>
    /// Auto-corrects common Claude mistakes in-place:
    /// - Normalize column widthFractions to sum to 1.0
    /// - Fix colSpan totals per row
    /// - Ensure colSpan/rowSpan >= 1
    /// </summary>
    public static void NormalizeInPlace(FormStructure structure)
    {
        foreach (var section in structure.Sections)
        {
            if (section.Table is null) continue;
            var table = section.Table;

            // 1. Normalize column fractions to sum to 1.0
            var totalFraction = table.Columns.Sum(c => c.WidthFraction);
            if (totalFraction > 0 && Math.Abs(totalFraction - 1.0) > 0.001)
            {
                var factor = 1.0 / totalFraction;
                foreach (var col in table.Columns)
                    col.WidthFraction *= factor;
            }

            // 2. Fix colSpan/rowSpan minimums
            foreach (var row in table.Rows)
            {
                foreach (var cell in row.Cells)
                {
                    if (cell.ColSpan < 1) cell.ColSpan = 1;
                    if (cell.RowSpan < 1) cell.RowSpan = 1;
                }
            }

            // 3. Fix colSpan totals per row
            var colCount = table.Columns.Count;
            foreach (var row in table.Rows)
            {
                var totalColSpan = row.Cells.Sum(c => c.ColSpan);

                if (totalColSpan < colCount)
                {
                    // Too few: expand the last cell's colSpan
                    var deficit = colCount - totalColSpan;
                    row.Cells[^1].ColSpan += deficit;
                }
                else if (totalColSpan > colCount)
                {
                    // Too many: shrink from the end
                    var excess = totalColSpan - colCount;
                    for (var i = row.Cells.Count - 1; i >= 0 && excess > 0; i--)
                    {
                        var canShrink = row.Cells[i].ColSpan - 1;
                        var shrink = Math.Min(canShrink, excess);
                        row.Cells[i].ColSpan -= shrink;
                        excess -= shrink;
                    }

                    // If still excess, remove trailing cells
                    while (row.Cells.Sum(c => c.ColSpan) > colCount && row.Cells.Count > 1)
                    {
                        row.Cells.RemoveAt(row.Cells.Count - 1);
                    }

                    // Final adjust on last cell
                    var remaining = colCount - row.Cells.Sum(c => c.ColSpan);
                    if (remaining > 0)
                        row.Cells[^1].ColSpan += remaining;
                    else if (remaining < 0 && row.Cells[^1].ColSpan > 1)
                        row.Cells[^1].ColSpan = Math.Max(1, row.Cells[^1].ColSpan + remaining);
                }
            }
        }
    }

    private static void ValidateSection(FormSection section, string path, List<string> errors)
    {
        var sectionType = (section.SectionType ?? "table").ToLowerInvariant();

        if (sectionType == "table")
        {
            ValidateTableSection(section, path, errors);
        }
        else if (sectionType is "freeform" or "keyvalue")
        {
            if (section.Elements is null or { Count: 0 })
            {
                errors.Add($"{path}: freeform/keyvalue section must have at least one element.");
            }
        }
    }

    private static void ValidateTableSection(FormSection section, string path, List<string> errors)
    {
        if (section.Table is null)
        {
            errors.Add($"{path}: table section must have a table definition.");
            return;
        }

        var table = section.Table;

        if (table.Columns.Count == 0)
        {
            errors.Add($"{path}.table: must have at least one column.");
            return;
        }

        if (table.Rows.Count == 0)
        {
            errors.Add($"{path}.table: must have at least one row.");
            return;
        }

        // After NormalizeInPlace, only check for truly broken rows
        for (var r = 0; r < table.Rows.Count; r++)
        {
            if (table.Rows[r].Cells.Count == 0)
            {
                errors.Add($"{path}.table.rows[{r}]: row must have at least one cell.");
            }
        }
    }
}

File: AutoEmply-server/src/main/java/health/autoemplyserver/service/FormStructureValidator.java
package health.autoemplyserver.service;

import health.autoemplyserver.model.FormSection;
import health.autoemplyserver.model.FormStructure;
import health.autoemplyserver.model.TableCell;
import health.autoemplyserver.model.TableDef;
import health.autoemplyserver.model.TableRow;
import java.util.ArrayList;
import java.util.List;
import org.springframework.stereotype.Component;

@Component
public class FormStructureValidator {

    public List<String> validate(FormStructure structure) {
        List<String> errors = new ArrayList<>();
        if (structure == null) {
            errors.add("FormStructure is null.");
            return errors;
        }
        if (structure.getSections() == null || structure.getSections().isEmpty()) {
            errors.add("FormStructure must have at least one section.");
            return errors;
        }

        normalizeInPlace(structure);

        for (int index = 0; index < structure.getSections().size(); index++) {
            validateSection(structure.getSections().get(index), "sections[" + index + "]", errors);
        }
        return errors;
    }

    public void normalizeInPlace(FormStructure structure) {
        for (FormSection section : structure.getSections()) {
            if (section.getTable() == null) {
                continue;
            }
            TableDef table = section.getTable();
            double totalFraction = table.getColumns().stream().mapToDouble(column -> column.getWidthFraction()).sum();
            if (totalFraction > 0 && Math.abs(totalFraction - 1.0d) > 0.001d) {
                for (var column : table.getColumns()) {
                    column.setWidthFraction(column.getWidthFraction() / totalFraction);
                }
            }
            for (TableRow row : table.getRows()) {
                for (TableCell cell : row.getCells()) {
                    if (cell.getColSpan() < 1) {
                        cell.setColSpan(1);
                    }
                    if (cell.getRowSpan() < 1) {
                        cell.setRowSpan(1);
                    }
                }
                int columnCount = table.getColumns().size();
                int totalColSpan = row.getCells().stream().mapToInt(TableCell::getColSpan).sum();
                if (totalColSpan < columnCount && !row.getCells().isEmpty()) {
                    row.getCells().getLast().setColSpan(row.getCells().getLast().getColSpan() + (columnCount - totalColSpan));
                } else if (totalColSpan > columnCount) {
                    int excess = totalColSpan - columnCount;
                    for (int idx = row.getCells().size() - 1; idx >= 0 && excess > 0; idx--) {
                        TableCell cell = row.getCells().get(idx);
                        int canShrink = cell.getColSpan() - 1;
                        int shrink = Math.min(canShrink, excess);
                        cell.setColSpan(cell.getColSpan() - shrink);
                        excess -= shrink;
                    }
                }
            }
        }
    }

    private void validateSection(FormSection section, String path, List<String> errors) {
        String sectionType = section.getSectionType() == null ? "table" : section.getSectionType().toLowerCase();
        if ("table".equals(sectionType)) {
            if (section.getTable() == null) {
                errors.add(path + ": table section must have a table definition.");
                return;
            }
            if (section.getTable().getColumns().isEmpty()) {
                errors.add(path + ".table: must have at least one column.");
            }
            if (section.getTable().getRows().isEmpty()) {
                errors.add(path + ".table: must have at least one row.");
            }
        } else if (("freeform".equals(sectionType) || "keyvalue".equals(sectionType))
            && (section.getElements() == null || section.getElements().isEmpty())) {
            errors.add(path + ": freeform/keyvalue section must have at least one element.");
        }
    }
}
